### PR TITLE
Allow ref to capture groups in mask replacement

### PR DESCRIPTION
**Pull request by NoisomePossum**
### What does this PR do?

Changes the replace function in logs mask sequences to expand references to capture groups (`$1`, `$2`, `${1}` etc.) allowing users to insert part of the matched content back into the replacement string.

### Motivation

Some customers have reached out recently asking for the ability to use backreferences to capture groups in the regex they specify for logs mask sequences. The use cases are generally as follows:
- user wants to use a certain string in a regex as a reference to the part they want to replace
- user wants to keep the reference string in the original log

Currently we only support inserting a raw string so customers have to hardcode the reference back in as in:

```
pattern: User=\w+
replace_placeholder: "User=[masked_user]"
```

However, sometimes the reference can be something variable as in:

```
Customer ID=1234
customer ID=1234
customer_id=1234
CustomerId=1234
```

In this case, customers are looking for something like:

```
pattern: ([^=]*=)\S+
replace_placeholder: "${1}[masked_id]"
```

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/logs/processor/processor_test.go [hunk 1]
```diff
@@ -113,6 +113,15 @@ func TestMask(t *testing.T) {
 	shouldProcess, redactedMessage = p.applyRedactingRules(newMessage([]byte("The credit card 4323124312341234 was used to buy some time"), &source, ""))
 	assert.Equal(t, true, shouldProcess)
 	assert.Equal(t, []byte("The credit card [masked_credit_card] was used to buy some time"), redactedMessage)
+
+	source = newSource("mask_sequences", "${1}[masked_value]", "([Dd]ata_?values=)\\S+")
```
Comment (ajacquemot) [inline pkg/logs/processor/processor_test.go]: can we please add test when there is nothing after the `=`please:
`data_values=` should return `data_values=` and not `data_values=[masked_value]`, right ?

**Review by ajacquemot — DISMISSED** [review]
Looks pretty cool, is @NBParis aware about this request ? Thanks for submitting this PR.

File: pkg/logs/processor/processor_test.go (hunk 1 above)
Comment (NoisomePossum) [reply pkg/logs/processor/processor_test.go]: Added. :) Indeed, for that test the regex fails to match so it doesn't perform the replacement.

I posted this in the logs T2 channel so @NBParis has seen it. 👍

**Review by ajacquemot — APPROVED** [review]

**Comment by NBParis** [comment]
That's indeed super cool. 
@NoisomePossum let's make sure we also update the documentation on this.

**Pull request merged**